api/client/identity.go

```diff
@@ -0,0 +1,138 @@
+limitations under the License.
+*/
+
+package client
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"os"
+
+	"github.com/gravitational/trace"
+)
+
+// IdentityFile represents the basic components of an identity file.
+type IdentityFile struct {
+	PrivateKey []byte
+	Certs      struct {
+		SSH []byte
+		TLS []byte
+	}
+	CACerts struct {
+		SSH [][]byte
+		TLS [][]byte
+	}
+}
+
+// TLS returns the identity file's associated tls config.
+func (idf *IdentityFile) TLS() (*tls.Config, error) {
+	cert, err := tls.X509KeyPair(idf.Certs.TLS, idf.PrivateKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	pool := x509.NewCertPool()
+	for i, caCerts := range idf.CACerts.TLS {
+		if !pool.AppendCertsFromPEM(caCerts) {
+			return nil, trace.BadParameter("identity file contains invalid TLS CA cert (#%v)", i+1)
+		}
+	}
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      pool,
+	}, nil
+}
+
+// DecodeIdentityFile attempts to break up the contents of an identity file
+// into its respective components.
+func DecodeIdentityFile(path string) (*IdentityFile, error) {
+	r, err := os.Open(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer r.Close()
+
+	scanner := bufio.NewScanner(r)
+	var ident IdentityFile
+	// Subslice of scanner's buffer pointing to current line
+	// with leading and trailing whitespace trimmed.
+	var line []byte
+	// Attempt to scan to the next line.
+	scanln := func() bool {
+		if !scanner.Scan() {
+			line = nil
+			return false
+		}
+		line = bytes.TrimSpace(scanner.Bytes())
+		return true
+	}
+	// Check if the current line starts with prefix `p`.
+	peekln := func(p string) bool {
+		return bytes.HasPrefix(line, []byte(p))
+	}
+	// Get an "owned" copy of the current line.
+	cloneln := func() []byte {
+		ln := make([]byte, len(line))
+		copy(ln, line)
+		return ln
+	}
+	// Scan through all lines of identity file.  Lines with a known prefix
+	// are copied out of the scanner's buffer.  All others are ignored.
+	for scanln() {
+		switch {
+		case peekln("ssh"):
+			ident.Certs.SSH = cloneln()
+		case peekln("@cert-authority"):
+			ident.CACerts.SSH = append(ident.CACerts.SSH, cloneln())
+		case peekln("-----BEGIN"):
+			// Current line marks the beginning of a PEM block.  Consume all
+			// lines until a corresponding END is found.
+			var pemBlock []byte
+			for {
+				pemBlock = append(pemBlock, line...)
+				pemBlock = append(pemBlock, '\n')
+				if peekln("-----END") {
+					break
+				}
+				if !scanln() {
+					// If scanner has terminated in the middle of a PEM block, either
+					// the reader encountered an error, or the PEM block is a fragment.
+					if err := scanner.Err(); err != nil {
+						return nil, trace.Wrap(err)
+					}
+					return nil, trace.BadParameter("invalid PEM block (fragment)")
+				}
+			}
+			// Decide where to place the pem block based on
+			// which pem blocks have already been found.
+			switch {
```

Comment:
Not sure I understand the strategy where a PEM block is placed - can you elaborate or describe the identity format for the reader?

It seems that this was moved from another location - feel free to disregard if it's not something you need addressed in this PR.

Reply:
I guess the PEM blocks are ordered this way, but I think there could be a better way to organize/clarify this as well. It doesn't fit into this PR though.